File: labs/2nd semestr/lab6/ts/Text.ts
import { MyString } from "./MyString";

export class Text {
    private lines: MyString[] = [];

    addLine(line: MyString): void {
        this.lines.push(line);
    }

    removeLine(index: number): void {
        if (index >= 0 && index < this.lines.length) {
            this.lines.splice(index, 1);
        }
    }

    clearText(): void {
        this.lines = [];
    }

    getLongestLine(): MyString | null {
        if (this.lines.length === 0) return null;
        return this.lines.reduce((longest, current) =>
            current.getContent().length > longest.getContent().length ? current : longest
        );
    }

    getDigitsPercentage(): number {
        const totalChars = this.lines.reduce((sum, line) => sum + line.getContent().length, 0);
        const totalDigits = this.lines.reduce((sum, line) => sum + line.countDigits(), 0);
        if (totalChars === 0) return 0;
        return (totalDigits / totalChars) * 100;
    }

    getTotalCharactersCount(): number {
        return this.lines.reduce((sum, line) => sum + line.getContent().length, 0);
    }
}

File: labs/2nd semestr/lab6/cs/Text.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace TextLib
{
    public class Text
    {
        private readonly List<MyString> lines = new();

        public void AddLine(MyString line) => lines.Add(line);

        public void RemoveLine(int index)
        {
            if (index >= 0 && index < lines.Count)
                lines.RemoveAt(index);
        }

        public void ClearText() => lines.Clear();

        public MyString? GetLongestLine()
        {
            return lines.OrderByDescending(l => l.GetContent().Length).FirstOrDefault();
        }

        public double GetDigitsPercentage()
        {
            int totalChars = lines.Sum(l => l.GetContent().Length);
            int totalDigits = lines.Sum(l => l.CountDigits());
            if (totalChars == 0) return 0;
            return (double)totalDigits / totalChars * 100;
        }

        public int GetTotalCharactersCount()
        {
            return lines.Sum(l => l.GetContent().Length);
        }
    }
}